### PR TITLE
Add RedisModule adapter

### PR DESCRIPTION
**Pull request by tezc**
Starting from this commit: https://github.com/redis/redis/pull/10001, redis (v7.0+) exposes its event loop from the module API. 
This PR adds an adapter to attach hiredis to redis' event loop. By doing that, a redis module can send commands to other redis instances via async hiredis.

### CODE REVIEW

File: examples/example-redismoduleapi.c [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+/*
+ * This example requires Redis 7.0 or above.
+ *
+ * 1- Compile this file as a shared library. Directory of "redismodule.h" must
+ *    be in the include path.
+ *       gcc -fPIC -shared -I../../redis/src/ -I.. example-redismoduleapi.c -o example-redismoduleapi.so
+ *
+ * 2- Load module:
+ *       redis-server --loadmodule ./example-redismoduleapi.so value
+ */
```
Comment (tezc) [inline examples/example-redismoduleapi.c:49]: I see we compile other adapter examples but I didn't add this to CMake or Makefile.
To compile this one, we need "redismodule.h" from redis source directory (maybe we may copy it here under examples/). 
Even if we compile this one, it won't be a standalone app as it is a module. 
So, I just left instructions here. Let me know if you prefer compile as part of CI 
@yossigo @michael-grunder

**Comment by tezc** [comment]
I just realize there is also `scheduleTimer` callback. We were using the adapter without it, I completely missed it. I'll implement it and then update the PR.

File: examples/example-redismoduleapi.c (hunk 1 above)
Comment (yossigo) [reply examples/example-redismoduleapi.c:49]: @tezc That makes sense. The value of merely compiling an example is limited, we need proper test coverage for all adapters, but that's an entirely different project.

File: examples/example-redismoduleapi.c [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+    if (ret != REDISMODULE_OK) {
+        printf("error module init \n");
+        return REDISMODULE_ERR;
+    }
+
+    if (!RedisModule_GetServerVersion ||
+        RedisModule_GetServerVersion() < 0x00070000) {
+        printf("Redis 7.0 or above is required! \n");
+        return REDISMODULE_ERR;
+    }
```
[the remark this answers is not in the document]
Comment (tezc) [reply examples/example-redismoduleapi.c:62]: Who is going to call that function? User? Maybe we can add those checks into `redisModuleAttach()` and return error on an older version?

[the remark this answers is not in the document]
Comment (tezc) [reply examples/example-redismoduleapi.c]: ok added it.

File: examples/example-redismoduleapi.c [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+        /* Let *c leak for now... */
+        printf("Error: %s\n", c->errstr);
+        return 1;
+    }
+
+    size_t len;
+    const char *val = RedisModule_StringPtrLen(argv[argc-1], &len);
+
+    RedisModuleCtx *module_ctx = RedisModule_GetDetachedThreadSafeContext(ctx);
+    redisModuleAttach(c, module_ctx);
```
Comment (tezc) [inline examples/example-redismoduleapi.c:86]: @yossigo Added timer callback as well. RM_CreateTimer() requires a RedisModuleCtx, so changed redisModuleAttach() prototype to accept a ctx. I assume we can at least pass detached context like this.

**Review by yossigo — APPROVED** [review]

**Review by michael-grunder — APPROVED** [review]

**Pull request merged**